Unable to apply subject-level RBAC bindings to cluster

**Describe the bug**
JulieOps is unable to apply subject-level RBAC bindings to the cluster. It fails with the following error:

```
[ERROR] 2021-05-27 16:26:25.979 [main] MDSApiClient - java.io.IOException: java.io.IOException: Something happened with the connection, response status code: 400 body: {"error_code":40002,"message":"Role binding update of resource-s
cope role without any resources"}
[ERROR] 2021-05-27 16:26:25.979 [main] MDSApiClient - java.io.IOException: java.io.IOException: Something happened with the connection, response status code: 400 body: {"error_code":40002,"message":"Role binding update of resource-s
cope role without any resources"}
[ERROR] 2021-05-27 16:26:25.979 [main] MDSApiClient - java.io.IOException: java.io.IOException: Something happened with the connection, response status code: 400 body: {"error_code":40002,"message":"Role binding update of resource-s
cope role without any resources"}
WARNING: sun.reflect.Reflection.getCallerClass is not supported. This will impact performance.
[ERROR] 2021-05-27 16:26:25.984 [main] ExecutionPlan - Something happen running action {
  "Operation" : "com.purbon.kafka.topology.actions.access.CreateBindings",
  "Bindings" : [ {
    "resourceType" : "CLUSTER",
    "resourceName" : "cluster",
    "host" : "*",
    "operation" : "ResourceOwner",
    "principal" : "User:user1",
    "pattern" : "LITERAL",
    "scope" : {
      "clusters" : {
        "kafka-cluster" : "...",
        "schema-registry-cluster" : "..."
      },
      "resources" : [ {
        "name" : "Subject:mysubject",
        "patternType" : "LITERAL",
        "resourceType" : "Subject"
      } ]
    }
  } ]
}
java.io.IOException: java.io.IOException: Something happened with the connection, response status code: 400 body: {"error_code":40002,"message":"Role binding update of resource-scope role without any resources"}
        at com.purbon.kafka.topology.clients.JulieHttpClient.doRequest(JulieHttpClient.java:125) ~[julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.clients.JulieHttpClient.doPost(JulieHttpClient.java:74) ~[julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.api.mds.MDSApiClient.bindRequest(MDSApiClient.java:88) ~[julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.roles.RBACProvider.createBindings(RBACProvider.java:29) ~[julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.actions.access.CreateBindings.execute(CreateBindings.java:28) ~[julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.actions.BaseAccessControlAction.run(BaseAccessControlAction.java:30) ~[julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.ExecutionPlan.execute(ExecutionPlan.java:112) ~[julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.ExecutionPlan.run(ExecutionPlan.java:90) [julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.JulieOps.run(JulieOps.java:242) [julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.JulieOps.run(JulieOps.java:259) [julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.CommandLineInterface.processTopology(CommandLineInterface.java:207) [julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.CommandLineInterface.run(CommandLineInterface.java:156) [julie-ops.jar.orig:?]
        at com.purbon.kafka.topology.CommandLineInterface.main(CommandLineInterface.java:146) [julie-ops.jar.orig:?]
Caused by: java.io.IOException: Something happened with the connection, response status code: 400 body: {"error_code":40002,"message":"Role binding update of resource-scope role without any resources"}
        at com.purbon.kafka.topology.clients.JulieHttpClient.doRequest(JulieHttpClient.java:114) ~[julie-ops.jar.orig:?]
        ... 12 more

```

**To Reproduce**

1. Create very simple topology file (no topics needed to reproduce the bug):

```
context: "mycontext"
tenant: "tenant-1"

projects:

  - name: "myapp"

    schemas:
      - principal: "User:user1"
        subjects:
          - "mysubject"

    topics: []
```

2. Run JulieOps with this topology against RBAC-enabled cluster 
(with JulieOps config `topology.builder.access.control.class = com.purbon.kafka.topology.roles.RBACProvider`)

3. Observe the exception as shown above.

**Expected behavior**
JulieOps sucessfully create subject-level RBAC binding based on the `schemas` block of the topology.

**Runtime (please complete the following information):**
 - OS: Windows
- JVM version: 11
 - Version 2.1.2

--- a/src/main/java/com/purbon/kafka/topology/api/mds/MDSApiClient.java
+++ b/src/main/java/com/purbon/kafka/topology/api/mds/MDSApiClient.java
@@ -70,22 +70,30 @@ public class MDSApiClient extends JulieHttpClient {
     return binding;
   }
 
-  public void bindRequest(TopologyAclBinding binding) throws IOException {
+  private boolean isBindingWithResources(TopologyAclBinding binding) {
+    return !binding.getScope().getResources().isEmpty();
+  }
 
+  MDSRequest buildRequest(TopologyAclBinding binding) {
     String url = binding.getPrincipal() + "/roles/" + binding.getOperation();
-    if (!binding.getResourceType().equals(ResourceType.CLUSTER.name())) {
-      url = url + "/bindings";
-    }
+    String jsonEntity;
 
+    if (isBindingWithResources(binding)) {
+      url = url + "/bindings";
+      jsonEntity = binding.getScope().asJson();
+    } else {
+      jsonEntity = binding.getScope().clustersAsJson();
+    }
+    LOGGER.debug("bind.entity: " + jsonEntity);
+
+    return new MDSRequest(url, jsonEntity);
+  }
+
+  public void bindRequest(TopologyAclBinding binding) throws IOException {
+    MDSRequest mdsRequest = buildRequest(binding);
     try {
-      String jsonEntity;
-      if (binding.getResourceType().equals(ResourceType.CLUSTER.name())) {
-        jsonEntity = binding.getScope().clustersAsJson();
-      } else {
-        jsonEntity = binding.getScope().asJson();
-      }
-      LOGGER.debug("bind.entity: " + jsonEntity);
-      doPost("/security/1.0/principals/" + url, jsonEntity);
+      LOGGER.debug("bind.entity: " + mdsRequest.getJsonEntity());
+      doPost("/security/1.0/principals/" + mdsRequest.getUrl(), mdsRequest.getJsonEntity());
     } catch (IOException e) {
       LOGGER.error(e);
       throw e;

--- a/src/main/java/com/purbon/kafka/topology/api/mds/MDSApiClient.java
+++ b/src/main/java/com/purbon/kafka/topology/api/mds/MDSApiClient.java
@@ -3,6 +3,7 @@ package com.purbon.kafka.topology.api.mds;
 import static com.purbon.kafka.topology.api.mds.RequestScope.RESOURCE_NAME;
 import static com.purbon.kafka.topology.api.mds.RequestScope.RESOURCE_PATTERN_TYPE;
 import static com.purbon.kafka.topology.api.mds.RequestScope.RESOURCE_TYPE;
+import static com.purbon.kafka.topology.roles.rbac.RBACPredefinedRoles.isClusterScopedRole;
 
 import com.purbon.kafka.topology.clients.JulieHttpClient;
 import com.purbon.kafka.topology.roles.TopologyAclBinding;
@@ -78,7 +79,7 @@ public class MDSApiClient extends JulieHttpClient {
     String url = binding.getPrincipal() + "/roles/" + binding.getOperation();
     String jsonEntity;
 
-    if (isBindingWithResources(binding)) {
+    if (isBindingWithResources(binding) && !isClusterScopedRole(binding.getOperation())) {
       url = url + "/bindings";
       jsonEntity = binding.getScope().asJson();
     } else {

--- a/src/main/java/com/purbon/kafka/topology/api/mds/MDSRequest.java
+++ b/src/main/java/com/purbon/kafka/topology/api/mds/MDSRequest.java
@@ -1,0 +1,19 @@
+package com.purbon.kafka.topology.api.mds;
+
+public class MDSRequest {
+  private final String url;
+  private final String jsonEntity;
+
+  public MDSRequest(String url, String jsonEntity) {
+    this.url = url;
+    this.jsonEntity = jsonEntity;
+  }
+
+  public String getUrl() {
+    return url;
+  }
+
+  public String getJsonEntity() {
+    return jsonEntity;
+  }
+}

--- a/src/main/java/com/purbon/kafka/topology/roles/rbac/RBACBindingsBuilder.java
+++ b/src/main/java/com/purbon/kafka/topology/roles/rbac/RBACBindingsBuilder.java
@@ -354,7 +354,12 @@ public class RBACBindingsBuilder implements BindingsBuilderProvider {
   @Override
   public List<TopologyAclBinding> setSchemaAuthorization(String principal, List<String> subjects) {
     return subjects.stream()
-        .map(subject -> apiClient.bind(principal, RESOURCE_OWNER).forSchemaSubject(subject).apply())
+        .map(
+            subject ->
+                apiClient
+                    .bind(principal, RESOURCE_OWNER)
+                    .forSchemaSubject(subject)
+                    .apply("Subject", subject))
         .filter(Objects::nonNull)
         .collect(Collectors.toList());
   }

--- a/src/main/java/com/purbon/kafka/topology/roles/rbac/RBACPredefinedRoles.java
+++ b/src/main/java/com/purbon/kafka/topology/roles/rbac/RBACPredefinedRoles.java
@@ -9,4 +9,10 @@ public class RBACPredefinedRoles {
 
   public static final String SECURITY_ADMIN = "SecurityAdmin";
   public static final String SYSTEM_ADMIN = "SystemAdmin";
+
+  public static boolean isClusterScopedRole(String role) {
+    return !DEVELOPER_READ.equals(role)
+            && !DEVELOPER_WRITE.equals(role)
+            && !RESOURCE_OWNER.equals(role);
+  }
 }

--- a/src/test/java/com/purbon/kafka/topology/api/mds/MDSApiClientTest.java
+++ b/src/test/java/com/purbon/kafka/topology/api/mds/MDSApiClientTest.java
@@ -1,0 +1,46 @@
+package com.purbon.kafka.topology.api.mds;
+
+import static com.purbon.kafka.topology.roles.rbac.RBACPredefinedRoles.DEVELOPER_READ;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+import com.purbon.kafka.topology.roles.TopologyAclBinding;
+import org.junit.Test;
+
+public class MDSApiClientTest {
+  MDSApiClient apiClient = new MDSApiClient("http://not_used:8090");
+
+  @Test
+  public void testBindSubjectRole() {
+    String principal = "User:foo";
+    String subject = "topic-value";
+
+    TopologyAclBinding binding =
+        apiClient
+            .bind(principal, DEVELOPER_READ)
+            .forSchemaSubject(subject)
+            .apply("Subject", subject);
+
+    MDSRequest mdsRequest = apiClient.buildRequest(binding);
+
+    assertThat(mdsRequest.getUrl()).isEqualTo("User:foo/roles/DeveloperRead/bindings");
+    assertThat(mdsRequest.getJsonEntity())
+        .isEqualTo(
+            "{\"resourcePatterns\":[{\"name\":\"Subject:topic-value\",\"patternType\":\"LITERAL\",\"resourceType\":\"Subject\"}],\"scope\":{\"clusters\":{\"kafka-cluster\":\"\",\"schema-registry-cluster\":\"\"}}}");
+  }
+
+  @Test
+  public void testBindSubjectRoleWithoutResourceType() {
+    String principal = "User:foo";
+    String subject = "topic-value";
+
+    TopologyAclBinding binding =
+        apiClient.bind(principal, DEVELOPER_READ).forSchemaSubject(subject).apply();
+
+    MDSRequest mdsRequest = apiClient.buildRequest(binding);
+
+    assertThat(mdsRequest.getUrl()).isEqualTo("User:foo/roles/DeveloperRead/bindings");
+    assertThat(mdsRequest.getJsonEntity())
+        .isEqualTo(
+            "{\"resourcePatterns\":[{\"name\":\"Subject:topic-value\",\"patternType\":\"LITERAL\",\"resourceType\":\"Subject\"}],\"scope\":{\"clusters\":{\"kafka-cluster\":\"\",\"schema-registry-cluster\":\"\"}}}");
+  }
+}

